File: packages/policy/clawdstrike-policy/src/plugins/manifest.ts
export type PluginTrustLevel = "trusted" | "untrusted";
export type PluginTrustSandbox = "node" | "wasm";
export type PluginGuardHandle =
  | "file_read"
  | "file_write"
  | "command_exec"
  | "network_egress"
  | "tool_call"
  | "patch_apply"
  | "secret_access"
  | "custom";

export interface PluginGuardManifestEntry {
  name: string;
  entrypoint: string;
  handles?: PluginGuardHandle[];
  configSchema?: string;
}

export interface PluginVersionCompatibility {
  minVersion?: string;
  maxVersion?: string;
}

export interface PluginCapabilities {
  network: boolean;
  subprocess: boolean;
  filesystem: {
    read: string[];
    write: boolean;
  };
  secrets: {
    access: boolean;
  };
}

export interface PluginResourceLimits {
  maxMemoryMb: number;
  maxCpuMs: number;
  maxTimeoutMs: number;
}

export interface PluginManifest {
  schema?: string;
  version: string;
  name: string;
  displayName?: string;
  description?: string;
  author?: string;
  license?: string;
  clawdstrike?: PluginVersionCompatibility;
  guards: PluginGuardManifestEntry[];
  capabilities: PluginCapabilities;
  resources: PluginResourceLimits;
  trust: {
    level: PluginTrustLevel;
    sandbox: PluginTrustSandbox;
  };
}

export function parsePluginManifest(value: unknown): PluginManifest {
  if (!isPlainObject(value)) {
    throw new Error("plugin manifest must be an object");
  }

  const version = value.version;
  if (typeof version !== "string" || version.trim() === "") {
    throw new Error("plugin manifest.version must be a non-empty string");
  }

  const name = value.name;
  if (typeof name !== "string" || name.trim() === "") {
    throw new Error("plugin manifest.name must be a non-empty string");
  }

  const trust = (value as any).trust;
  if (!isPlainObject(trust)) {
    throw new Error("plugin manifest.trust must be an object");
  }
  const level = (trust as any).level;
  if (level !== "trusted" && level !== "untrusted") {
    throw new Error('plugin manifest.trust.level must be "trusted" or "untrusted"');
  }
  const sandboxRaw = (trust as any).sandbox;
  if (sandboxRaw !== undefined && sandboxRaw !== "node" && sandboxRaw !== "wasm") {
    throw new Error('plugin manifest.trust.sandbox must be "node" or "wasm"');
  }
  const sandbox: PluginTrustSandbox = sandboxRaw ?? "node";

  const compatibility = parseCompatibility((value as any).clawdstrike);
  const capabilities = parseCapabilities((value as any).capabilities);
  const resources = parseResources((value as any).resources);

  const guards = (value as any).guards;
  if (!Array.isArray(guards) || guards.length === 0) {
    throw new Error("plugin manifest.guards must be a non-empty array");
  }

  const guardNames = new Set<string>();
  const parsedGuards: PluginGuardManifestEntry[] = [];
  for (let i = 0; i < guards.length; i++) {
    const g = guards[i];
    const base = `plugin manifest.guards[${i}]`;
    if (!isPlainObject(g)) {
      throw new Error(`${base} must be an object`);
    }

    const guardName = (g as any).name;
    if (typeof guardName !== "string" || guardName.trim() === "") {
      throw new Error(`${base}.name must be a non-empty string`);
    }
    if (guardNames.has(guardName)) {
      throw new Error(`${base}.name duplicates guard: ${guardName}`);
    }
    guardNames.add(guardName);

    const entrypoint = (g as any).entrypoint;
    if (typeof entrypoint !== "string" || entrypoint.trim() === "") {
      throw new Error(`${base}.entrypoint must be a non-empty string`);
    }

    const handlesRaw = (g as any).handles;
    const handles = parseHandles(handlesRaw, `${base}.handles`);

    const configSchema = (g as any).configSchema;
    if (
      configSchema !== undefined &&
      (typeof configSchema !== "string" || configSchema.trim() === "")
    ) {
      throw new Error(`${base}.configSchema must be a non-empty string when provided`);
    }

    parsedGuards.push({
      name: guardName,
      entrypoint,
      handles,
      configSchema,
    });
  }

  return {
    schema:
      typeof (value as any).$schema === "string" && (value as any).$schema.trim() !== ""
        ? String((value as any).$schema)
        : undefined,
    version,
    name,
    displayName: parseOptionalString((value as any).displayName, "plugin manifest.displayName"),
    description: parseOptionalString((value as any).description, "plugin manifest.description"),
    author: parseOptionalString((value as any).author, "plugin manifest.author"),
    license: parseOptionalString((value as any).license, "plugin manifest.license"),
    clawdstrike: compatibility,
    guards: parsedGuards,
    capabilities,
    resources,
    trust: { level, sandbox },
  };
}

function parseHandles(value: unknown, base: string): PluginGuardHandle[] | undefined {
  if (value === undefined) {
    return undefined;
  }
  if (!Array.isArray(value)) {
    throw new Error(`${base} must be an array when provided`);
  }

  const out: PluginGuardHandle[] = [];
  for (let i = 0; i < value.length; i++) {
    const v = value[i];
    if (!isPluginGuardHandle(v)) {
      throw new Error(`${base}[${i}] must be a valid handle event type`);
    }
    out.push(v);
  }
  return out;
}

function parseCompatibility(value: unknown): PluginVersionCompatibility | undefined {
  if (value === undefined) {
    return undefined;
  }
  if (!isPlainObject(value)) {
    throw new Error("plugin manifest.clawdstrike must be an object when provided");
  }

  const minVersion = parseOptionalString(
    value.minVersion,
    "plugin manifest.clawdstrike.minVersion",
  );
  if (minVersion !== undefined && !isStrictSemver(minVersion)) {
    throw new Error("plugin manifest.clawdstrike.minVersion must be strict semver (x.y.z)");
  }

  const maxVersion = parseOptionalString(
    value.maxVersion,
    "plugin manifest.clawdstrike.maxVersion",
  );
  if (maxVersion !== undefined && !isSemverRange(maxVersion)) {
    throw new Error(
      "plugin manifest.clawdstrike.maxVersion must be semver or wildcard range (e.g. 1.x)",
    );
  }

  return { minVersion, maxVersion };
}

function parseCapabilities(value: unknown): PluginCapabilities {
  if (value === undefined) {
    return defaultCapabilities();
  }
  if (!isPlainObject(value)) {
    throw new Error("plugin manifest.capabilities must be an object when provided");
  }

  const network =
    parseOptionalBoolean(value.network, "plugin manifest.capabilities.network") ?? false;
  const subprocess =
    parseOptionalBoolean(value.subprocess, "plugin manifest.capabilities.subprocess") ?? false;

  const filesystemRaw = value.filesystem;
  let filesystem = { read: [] as string[], write: false };
  if (filesystemRaw !== undefined) {
    if (!isPlainObject(filesystemRaw)) {
      throw new Error("plugin manifest.capabilities.filesystem must be an object when provided");
    }

    const readRaw = filesystemRaw.read;
    let read: string[] = [];
    if (readRaw !== undefined) {
      if (readRaw === false) {
        read = [];
      } else if (Array.isArray(readRaw)) {
        read = readRaw.map((v, i) => {
          if (typeof v !== "string" || v.trim() === "") {
            throw new Error(
              `plugin manifest.capabilities.filesystem.read[${i}] must be a non-empty string`,
            );
          }
          return v;
        });
      } else {
        throw new Error(
          "plugin manifest.capabilities.filesystem.read must be false or an array of strings",
        );
      }
    }

    const write =
      parseOptionalBoolean(filesystemRaw.write, "plugin manifest.capabilities.filesystem.write") ??
      false;
    filesystem = { read, write };
  }

  const secretsRaw = value.secrets;
  let secretsAccess = false;
  if (secretsRaw !== undefined) {
    if (typeof secretsRaw === "boolean") {
      secretsAccess = secretsRaw;
    } else if (isPlainObject(secretsRaw)) {
      secretsAccess =
        parseOptionalBoolean(secretsRaw.access, "plugin manifest.capabilities.secrets.access") ??
        false;
    } else {
      throw new Error("plugin manifest.capabilities.secrets must be a boolean or object");
    }
  }

  return {
    network,
    subprocess,
    filesystem,
    secrets: {
      access: secretsAccess,
    },
  };
}

function parseResources(value: unknown): PluginResourceLimits {
  const defaults = defaultResources();
  if (value === undefined) {
    return defaults;
  }
  if (!isPlainObject(value)) {
    throw new Error("plugin manifest.resources must be an object when provided");
  }

  const maxMemoryMb =
    parseOptionalPositiveInt(value.maxMemoryMb, "plugin manifest.resources.maxMemoryMb") ??
    defaults.maxMemoryMb;
  const maxCpuMs =
    parseOptionalPositiveInt(value.maxCpuMs, "plugin manifest.resources.maxCpuMs") ??
    defaults.maxCpuMs;
  const maxTimeoutMs =
    parseOptionalPositiveInt(value.maxTimeoutMs, "plugin manifest.resources.maxTimeoutMs") ??
    defaults.maxTimeoutMs;

  return {
    maxMemoryMb,
    maxCpuMs,
    maxTimeoutMs,
  };
}

function parseOptionalString(value: unknown, field: string): string | undefined {
  if (value === undefined) {
    return undefined;
  }
  if (typeof value !== "string" || value.trim() === "") {
    throw new Error(`${field} must be a non-empty string when provided`);
  }
  return value;
}

function parseOptionalBoolean(value: unknown, field: string): boolean | undefined {
  if (value === undefined) {
    return undefined;
  }
  if (typeof value !== "boolean") {
    throw new Error(`${field} must be a boolean when provided`);
  }
  return value;
}

function parseOptionalPositiveInt(value: unknown, field: string): number | undefined {
  if (value === undefined) {
    return undefined;
  }
  if (typeof value !== "number" || !Number.isInteger(value) || value < 1) {
    throw new Error(`${field} must be a positive integer when provided`);
  }
  return value;
}

function isPluginGuardHandle(value: unknown): value is PluginGuardHandle {
  return (
    value === "file_read" ||
    value === "file_write" ||
    value === "command_exec" ||
    value === "network_egress" ||
    value === "tool_call" ||
    value === "patch_apply" ||
    value === "secret_access" ||
    value === "custom"
  );
}

function isStrictSemver(value: string): boolean {
  return /^(\d+)\.(\d+)\.(\d+)$/.test(value);
}

function isSemverRange(value: string): boolean {
  return (
    /^(\d+)\.(\d+)\.(\d+)$/.test(value) ||
    /^(\d+)\.x$/.test(value) ||
    /^(\d+)\.(\d+)\.x$/.test(value)
  );
}

function defaultCapabilities(): PluginCapabilities {
  return {
    network: false,
    subprocess: false,
    filesystem: {
      read: [],
      write: false,
    },
    secrets: {
      access: false,
    },
  };
}

function defaultResources(): PluginResourceLimits {
  return {
    maxMemoryMb: 64,
    maxCpuMs: 100,
    maxTimeoutMs: 5000,
  };
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: crates/libs/clawdstrike/src/plugins/manifest.rs
use std::collections::HashSet;

use serde::{Deserialize, Serialize};

use crate::error::{Error, Result};
use crate::semver_utils::{is_strict_semver, parse_strict_semver};

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct PluginManifest {
    pub plugin: PluginMetadata,
    #[serde(default)]
    pub clawdstrike: Option<PluginClawdstrikeCompatibility>,
    #[serde(default)]
    pub guards: Vec<PluginGuardManifestEntry>,
    #[serde(default)]
    pub capabilities: PluginCapabilities,
    #[serde(default)]
    pub resources: PluginResourceLimits,
    #[serde(default)]
    pub trust: PluginTrust,
}

impl PluginManifest {
    pub fn validate(&self) -> Result<()> {
        if self.plugin.version.trim().is_empty() {
            return Err(Error::ConfigError(
                "plugin manifest.plugin.version must be a non-empty string".to_string(),
            ));
        }
        if self.plugin.name.trim().is_empty() {
            return Err(Error::ConfigError(
                "plugin manifest.plugin.name must be a non-empty string".to_string(),
            ));
        }
        if self.guards.is_empty() {
            return Err(Error::ConfigError(
                "plugin manifest.guards must be a non-empty array".to_string(),
            ));
        }
        if self.resources.max_memory_mb < 1
            || self.resources.max_cpu_ms < 1
            || self.resources.max_timeout_ms < 1
        {
            return Err(Error::ConfigError(
                "plugin manifest.resources values must be positive integers".to_string(),
            ));
        }

        if let Some(compat) = &self.clawdstrike {
            if let Some(min) = &compat.min_version {
                if !is_strict_semver(min) {
                    return Err(Error::ConfigError(
                        "plugin manifest.clawdstrike.min_version must be strict semver (x.y.z)"
                            .to_string(),
                    ));
                }
            }

            if let Some(max) = &compat.max_version {
                if !is_semver_range(max) {
                    return Err(Error::ConfigError(
                        "plugin manifest.clawdstrike.max_version must be semver or wildcard range (e.g. 1.x)"
                            .to_string(),
                    ));
                }
            }
        }

        let mut names = HashSet::new();
        for (idx, guard) in self.guards.iter().enumerate() {
            if guard.name.trim().is_empty() {
                return Err(Error::ConfigError(format!(
                    "plugin manifest.guards[{idx}].name must be a non-empty string"
                )));
            }
            if !names.insert(guard.name.clone()) {
                return Err(Error::ConfigError(format!(
                    "plugin manifest.guards[{idx}].name duplicates guard: {}",
                    guard.name
                )));
            }
            if let Some(entrypoint) = &guard.entrypoint {
                if entrypoint.trim().is_empty() {
                    return Err(Error::ConfigError(format!(
                        "plugin manifest.guards[{idx}].entrypoint must be a non-empty string when provided"
                    )));
                }
            }
        }

        if self.trust.level == PluginTrustLevel::Untrusted {
            if self.capabilities.subprocess {
                return Err(Error::ConfigError(format!(
                    "untrusted plugin {} cannot request subprocess capability",
                    self.plugin.name
                )));
            }
            if self.capabilities.filesystem.write {
                return Err(Error::ConfigError(format!(
                    "untrusted plugin {} cannot request filesystem write capability",
                    self.plugin.name
                )));
            }
            if self.capabilities.secrets.access {
                return Err(Error::ConfigError(format!(
                    "untrusted plugin {} cannot request secrets access capability",
                    self.plugin.name
                )));
            }
        }

        Ok(())
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct PluginMetadata {
    pub version: String,
    pub name: String,
    #[serde(default)]
    pub display_name: Option<String>,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub author: Option<String>,
    #[serde(default)]
    pub license: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct PluginClawdstrikeCompatibility {
    #[serde(default)]
    pub min_version: Option<String>,
    #[serde(default)]
    pub max_version: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct PluginGuardManifestEntry {
    pub name: String,
    #[serde(default)]
    pub display_name: Option<String>,
    #[serde(default)]
    pub entrypoint: Option<String>,
    #[serde(default)]
    pub handles: Vec<String>,
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct PluginCapabilities {
    #[serde(default)]
    pub network: bool,
    #[serde(default)]
    pub subprocess: bool,
    #[serde(default)]
    pub filesystem: PluginFilesystemCapabilities,
    #[serde(default)]
    pub secrets: PluginSecretsCapabilities,
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct PluginFilesystemCapabilities {
    #[serde(default)]
    pub read: Vec<String>,
    #[serde(default)]
    pub write: bool,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(deny_unknown_fields)]
pub struct PluginSecretsCapabilities {
    #[serde(default)]
    pub access: bool,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum PluginTrustLevel {
    Trusted,
    Untrusted,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum PluginSandbox {
    Native,
    Wasm,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct PluginTrust {
    pub level: PluginTrustLevel,
    pub sandbox: PluginSandbox,
}

impl Default for PluginTrust {
    fn default() -> Self {
        Self {
            level: PluginTrustLevel::Untrusted,
            sandbox: PluginSandbox::Wasm,
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct PluginResourceLimits {
    #[serde(default = "default_max_memory_mb")]
    pub max_memory_mb: u32,
    #[serde(default = "default_max_cpu_ms")]
    pub max_cpu_ms: u32,
    #[serde(default = "default_max_timeout_ms")]
    pub max_timeout_ms: u32,
}

impl Default for PluginResourceLimits {
    fn default() -> Self {
        Self {
            max_memory_mb: default_max_memory_mb(),
            max_cpu_ms: default_max_cpu_ms(),
            max_timeout_ms: default_max_timeout_ms(),
        }
    }
}

const fn default_max_memory_mb() -> u32 {
    64
}

const fn default_max_cpu_ms() -> u32 {
    100
}

const fn default_max_timeout_ms() -> u32 {
    5000
}

pub fn parse_plugin_manifest_toml(content: &str) -> Result<PluginManifest> {
    let manifest: PluginManifest = toml::from_str(content)
        .map_err(|e| Error::ConfigError(format!("failed to parse plugin manifest TOML: {e}")))?;
    manifest.validate()?;
    Ok(manifest)
}

fn is_semver_range(value: &str) -> bool {
    parse_strict_semver(value).is_some()
        || parse_major_wildcard(value).is_some()
        || parse_minor_wildcard(value).is_some()
}

fn parse_major_wildcard(value: &str) -> Option<u32> {
    let mut parts = value.split('.');
    let major = parts.next()?.parse::<u32>().ok()?;
    let mid = parts.next()?;
    if mid != "x" || parts.next().is_some() {
        return None;
    }
    Some(major)
}

fn parse_minor_wildcard(value: &str) -> Option<(u32, u32)> {
    let mut parts = value.split('.');
    let major = parts.next()?.parse::<u32>().ok()?;
    let minor = parts.next()?.parse::<u32>().ok()?;
    let patch = parts.next()?;
    if patch != "x" || parts.next().is_some() {
        return None;
    }
    Some((major, minor))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_minimal_manifest_and_defaults() {
        let raw = r#"
[plugin]
version = "1.0.0"
name = "acme-guard"

[[guards]]
name = "acme.deny"

[trust]
level = "trusted"
sandbox = "native"
"#;

        let manifest = parse_plugin_manifest_toml(raw).expect("manifest parsed");
        assert_eq!(manifest.plugin.name, "acme-guard");
        assert_eq!(manifest.resources.max_memory_mb, 64);
        assert!(!manifest.capabilities.network);
        assert!(!manifest.capabilities.filesystem.write);
    }

    #[test]
    fn rejects_duplicate_guard_names() {
        let raw = r#"
[plugin]
version = "1.0.0"
name = "acme-guard"

[[guards]]
name = "acme.deny"

[[guards]]
name = "acme.deny"

[trust]
level = "trusted"
sandbox = "native"
"#;

        let err = parse_plugin_manifest_toml(raw).expect_err("duplicate should fail");
        assert!(err.to_string().contains("duplicates guard"));
    }

    #[test]
    fn rejects_untrusted_high_risk_caps() {
        let raw = r#"
[plugin]
version = "1.0.0"
name = "acme-untrusted"

[[guards]]
name = "acme.guard"

[capabilities]
subprocess = true

[trust]
level = "untrusted"
sandbox = "wasm"
"#;

        let err = parse_plugin_manifest_toml(raw).expect_err("should reject subprocess");
        assert!(err
            .to_string()
            .contains("cannot request subprocess capability"));
    }
}
